Add initial docs and some diagrams (#79)
weaveworks/profiles: README.md

# profiles
Gitops native package management

<!--
To update the TOC, install https://github.com/kubernetes-sigs/mdtoc
and run: mdtoc -inplace README.md
-->

 <!-- toc -->
- [Getting started](#getting-started)
  - [Local environment using <a href="https://kind.sigs.k8s.io/">Kind</a>](#local-environment-using-kind)
  - [Local development](#local-development)

>   - [Installing Profiles](#installing-profiles)

Reviewer: feels like some sort of intro & tldr is missing, going straight into local-env setup is jumping the gun I think

Author: yeh i will add more in another PR

- [Terminology](#terminology)
  - [Profile](#profile)
  - [Catalog](#catalog)
  - [Profile Catalog Source](#profile-catalog-source)
  - [Profile Subscription](#profile-subscription)
  - [Profile Controller](#profile-controller)
  - [Profile Catalog Source Controller](#profile-catalog-source-controller)
- [Current Architecture](#current-architecture)
  - [Catalogs and Sources](#catalogs-and-sources)
  - [Profile Subscriptions](#profile-subscriptions)
- [Roadmap](#roadmap)
  - [Profiles](#profiles)
  - [Catalogs](#catalogs)
<!-- /toc -->

## Getting started

### Local environment using [Kind](https://kind.sigs.k8s.io/)

1. Set up local environment: `make local-env`.

    This will start a local `kind` cluster and installs

>     the `profilesubscription`, `source` and `helm` controllers.

Reviewer: ```suggestion
    the `profiles` and `flux` components.
```

maybe `gitops toolkit`?

Author: flux is part of gotk, not all of it

1. Deploy an example catalog source `kubectl apply -f examples/profile-catalog-source.yaml`

> 1. Expose the catalog API: `kubectl -n profiles-system port-forward $(kubectl get pods -A | awk '/profiles-controller-manager/ {print $2}') 8000:8000`

Reviewer: imminently going to be outdated :smile:

### Local development

> 1. Tests can be run with `make test`. (note: the previous env setup needs to be done first.)

Reviewer: ```suggestion
1. Tests can be run with `make test`.
```

env only needed for acceptance I think?

Author: oh yeh, and the acceptance command will do local env anyway

1. Acceptance tests can be run with `make acceptance`.

1. For further commands, run `make help`.

### Installing Profiles

1. Profiles can be installed using [pctl](https://github.com/weaveworks/pctl).

## Terminology

### Profile

A Profile is a "package" of Kubernetes deployable objects, known as Artifacts, and configurable values.

> Artifacts are one of: Helm Chart; Helm Release; raw yaml; Kustomize patch.

Reviewer: ```suggestion
Artifacts are one of: Helm Chart; Helm Release; raw yaml; Kustomize patch; another profile.
```

For an example, see the [nginx-profile](https://github.com/weaveworks/nginx-profile).

### Catalog

A Catalog is an in-memory cache of Profiles. There is one Catalog per running [Profile Controller](#profile-controller).

> The Catalog is queryable via an API which runs alongside the Profiles Controller.

Reviewer: ```suggestion
The Catalog is queryable via [pctl](https://github.com/weaveworks/pctl) or the API directly which runs alongside the Profiles Controller.
```

Artifact objects are processed by Flux components, with the health of Subscription children
reflected in the `ProfileSubscriptionStatus`.

```go
// ProfileSubscriptionSpec defines the desired state of a ProfileSubscription
type ProfileSubscriptionSpec struct {
	// ProfileURL is a fully qualified URL to a profile repo
	ProfileURL string `json:"profileURL,omitempty"`
	// Branch is the git repo branch containing the profile definition (default: main)
	// +kubebuilder:default:=main
	// +optional
	Branch string `json:"branch,omitempty"`

	// Values holds the values for the Helm chart specified in the first artifact
	// +optional
	Values *apiextensionsv1.JSON `json:"values,omitempty"`

	// ValuesFrom holds references to resources containing values for the Helm chart specified in the first artifact
	// +optional
	ValuesFrom []helmv2.ValuesReference `json:"valuesFrom,omitempty"`
}

// ProfileSubscriptionStatus defines the observed state of ProfileSubscription
type ProfileSubscriptionStatus struct {
	// Conditions holds the conditions for the ProfileSubscription
	// +optional
	Conditions []metav1.Condition `json:"conditions,omitempty"`

> }

Reviewer: I recall theres someway to automate generating an open-api specs for CRDs? Will automatically keep this up to date

Author: oh yeh i meant to look that up

Reviewer: I'll leave it up to you if you want to do it as part of this PR

Author: can't find anything on how to do this now, so will just add it later

```

> Profiles can only be installed if they are listed in a [Catalog Source](#profile-catalog-source).

Reviewer: is that true?

Author: currently, yes i think

but will change

Reviewer: where does it get enforced? is it just in pctl? I think I can create a profilesub manually and it would work for any profile

Author: oh i mean if done the "official" way by pctl. i will clarify

### Profile Controller

> The Profile Controller reconciles `ProfileSubscription` resources.

Reviewer: ```suggestion
The Profile Controller reconciles `ProfileSubscription` resources. See architecture diagrams below for what the reconciliation process does
```
